[Vending machine step2] 자판기 기본 동작

1Consumption opened the pull request:
* 인터페이스 메소드 관련해서 게터와 세터가 많은 것 같다고 생각도비니다.
* 브랜드 이름이나, 음료의 이름 같은 경우는 열거형으로 어느 정도 제한을 줘야한다고 생각이 됩니다만, 이는 추후 스텝에서 필요 시 수정하도록 하겠습니다.
* 최대한 내부 프로퍼티에 직접 접근하는 것을 지양하고, 동작을 넘겨 해당 클래스에서 처리하도록 노력하였습니다.
* 위와 같은 맥락으로 유닛 테스트 클래스에도 동일하게 적용해봤습니다.

1Consumption committed:
date를 사용자가 원하는 형식으로 반환 해주는 클래스

1Consumption committed:
VendingMachin 클래스 변경

동작을 넘겨서 Beverages가 수행하게끔 수정함

1Consumption committed:
디렉토리 세분화

1Consumption committed:
Milk 클래스 생성

농장 코드를 프로퍼티로 가짐
farmCode에 대한 인터페이스 메소드 가지고 있음

1Consumption committed:
속성 추가

유통기한, 칼로리 속성 추가

1Consumption committed:
Soda 클래스 추가

Beverage클래스를 상속받음
zeroCalorie 프로퍼티를 가지고 있음

1Consumption committed:
Coffee 클래스 추가

Beverage 클래스를 상속 받음
뜨거운지 여부에 대한 hot속성에 대한 인터페이스
디카페인인지 여부에 대한 decaffeinated 속성에 대한 인터페이스 가지고 있음

1Consumption committed:
StrawberryMilk 클래스 추가

딸기과즙 함유량을 표현하는 juiceContent 속성 가짐
매개변수로 주어진 함유량보다 높은지 낮은지를 판단하는 인터페이스 가짐

1Consumption committed:
띄어쓰기 및 접근자 수정

1Consumption committed:
ChocolateMilk 클래스 생성

패키지 종류를 Package enum으로 선언하여 package 프로퍼티가 가지고 있음
패키지 종류가 같은지를 검사하는 인터페이스 가짐

1Consumption committed:
Coke 클래스 추가

강한 탄산이 들어있는지 여부를 나타내는 strongCarbonicAcid 프로퍼티를 가짐
strongCarbonicAcid 프로퍼티에 대한 인터페이스 메소드 가지고 있음

1Consumption committed:
Cider 클래스 추가

맛을 표현하는 Taste enum을 가짐
taste 속성이 맛을 가짐
taste가 같은지 검사하는 인터페이스 메소드 있음

1Consumption committed:
TOP 클래스 추가

원산지 정보를 가지는 beanOrigin를 속성을 가짐
원산지가 같은지 여부를 판단하는 인터페이스 메소드 있음

1Consumption committed:
Georgia 클래스 추가

블랙 커피인지 아닌지를 표시하는 black 프로퍼티를 가짐

1Consumption committed:
balance 프로퍼티 추가

잔액에 대한 값을가짐
addBalance 메소드가 인터페이스 역할을 함

1Consumption committed:
VendingMachine 클래스 기능 추가

Beverage 객체들의 가격과 현재 잔액을 비교하기 위해 Beverage 클래스에 canBuy() 메소드를 만들어줌
vendingMachine 에서 가지고 있는 Beverage를 순회하면서 구매 가능한 음료 리스트를 넘겨주는 purchasableBeverages() 메소드 작성

1Consumption committed:
음료수 구매 기능 추가

VendingMachine 클래스에 purchase 메소드를 통해 해당 기능 실행 가능
Beverage 객체끼리 비교할 수 있도록 Equatable 프로토콜 채용

1Consumption committed:
[기능 추가] 잔액 반환

balance에 대한 인터페이스 메소드 만들어 줌.

1Consumption committed:
[기능 추가] 음료수를 종류별로 리턴

Beverages 클래스에서 딕셔너리를 리턴함
제조일자는 달라도 종류가 같으면 같은 음료수이기 때문에 음료수를 표현하는 방식에서 제조일자를 뺌

1Consumption committed:
[기능 추가] 유통기한 만료 음료수 반환

Beverage 클래스 내 isSafe() 메소드를 활용하여 값이 false 인 경우 배열에 담아 리턴

1Consumption committed:
[기능 수정] hot 프로퍼티를 Beverage 클래스로 옮김

1Consumption committed:
[기능 추가] 뜨거운 음료 배열 리턴

Beverage 클래스 내 isHot() 메소드의 결과가 true면 배열에 담아 리턴

1Consumption committed:
[기능 추가] 구매 내역 리턴

구매한 음료의 정보를 가지는 Beverages클래스인 purchaseHistory 속성을 선언
purchasedBeverages 메소드에서 구매 내역을 리턴

1Consumption committed:
[기능 추가] VendingMachine 클래스 유닛 테스트

1Consumption committed:
[기능 추가] Beverages 클래스 유닛 테스트

1Consumption committed:
[기능 추가] Beverage 클래스 유닛 테스트

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -0,0 +1,59 @@
+    
+    func getPrice() -> Int {
+        return price
+    }
```
getter가 꼭 필요한 경우라면 private(set)도 고려해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -0,0 +1,59 @@
+
+extension Beverage: CustomStringConvertible {
+    var description: String {
+        return "\(brand), \(capacity)ml, \(price)원, \(name)"
```
ml 이나 원 같은 단위를 붙이는 건 출력하기 위해서 일텐데
음료 객체가 보관하는 값과 단위를 붙여서 처리하지 않는 게 좋습니다.
왜냐하면 항상 이런 문자열로 출력한다는 보장이 없거든요. 
앱에서는 뷰에 어떤 화면 구성인지, 지역이나 언어 설정에 따라 로컬라이즈를 하기도 하고 
값을 표시하기 위한 형식은 모델 자체가 구현하기 보다 출력하기 전에 다루기 때문입니다. 참고하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverages.swift:
```diff
@@ -0,0 +1,46 @@
+    }
+    
+    func removeBeverage(beverage: Beverage) {
+        for beverageIndex in 0 ..< beverages.count {
```
메소드 이름에서도 addBeverage(bevergae:) , removeBeverage(beverage:) 보다는
add(beverage:) , remove(beverage:) 로 표현하는 게 읽기 편할 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverages.swift:
```diff
@@ -0,0 +1,46 @@
+            beverageDic["\(beverage)"] = beverages.filter{beverage == $0}.count
+        }
+        
+        return beverageDic
```
메소드에 타입이름을 붙이지 말고, 그 역할이 뭔지 동작이 뭔지 설명할 수 있는게 좋습니다.
getDictionary가 의도를 표현하기는 부족하니까요. 
스위프트 API 디자인 가이드 문서를 읽어보세요. 
변수명을 붙일 때도 타입을 붙이기 보다는 역할을 붙이도록 권장하고 있습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverages.swift:
```diff
@@ -0,0 +1,46 @@
+        beverages.forEach{
+            let beverage = $0
+            beverageDic["\(beverage)"] = beverages.filter{beverage == $0}.count
+        }
```
여기 구현 부분을 reduce를 활용할 수 있으니까 reduce로 바꿔보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Coke.swift:
```diff
@@ -0,0 +1,22 @@
+import Foundation
+
+class Coke: Soda {
+    private var strongCarbonicAcid: Bool
```
타입을 bool 로 만들면 strong 인지 아닌지는 어디서 어떻게 판단하게 될까요?

1Consumption replied:
콜라 종류에 강한 탄산과 기본이 있어서 그 부분을 표현하려고 Bool 타입으로 두었습니다.
하지만 해당 프로퍼티를 Double 타입으로 설정하여 내부 메소드에서 강한 탄산 여부를 판단하게끔 하는 방법을 고려해보겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -0,0 +1,59 @@
+    private var manufacturingDate: Date
+    private var expirationDate: Date
+    private var calorie: Double
+    private var hot: Bool
```
이런 값을 모델로 표현할 때는 hot : Bool 로 하기 보다는 
온도를 숫자로 표현하고 기준보다 뜨거운지 차가운지 표현하는 게 좋습니다. 
Bool 타입은 나누기 매우 명확한 경우에만 쓰세요

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+    
+    func isSafe() -> Bool {
+        return CustomDateFormatter.convertDate(date: expirationDate) > CustomDateFormatter.convertDate(date : Date())
```
Date 타입끼리 비교할 수 있지 않던가요? 왜 변환해서 비교하게 되나요?

1Consumption replied:
아... 그 부분은 미처 몰랐습니다. Date 타입끼리 비교하도록 하겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -0,0 +1,97 @@
+            purchased.append("\($0)")
+        }
+        
+        return purchased
```
위에 클로저가 있으면 이 메소드는 필요없을 수도 있겠네요.

godrm reviewed (commented):
클래스 계층 구조를 잘 나누셨네요. 다음 피드백을 확인해보세요.

1Consumption committed:
[기능 수정] getter 제거

balance에 대한 인터페이스 메소드인 getBalance 메소드를 제거하고 접근자를 private(set)로 수정함

1Consumption committed:
[기능 수정] description 수정

단위 제거

1Consumption committed:
[기능 수정] 메소드 이름 수정

중복되는 이름 변경

1Consumption committed:
[기능 수정] 메소드 이름 수정

타입을 포함하지 않고 해당 메소드의 역할을 이름으로 설정함

1Consumption committed:
[기능 수정] 고차함수 적용

forEach문을 reduce로 대체

1Consumption committed:
[기능 수정] 각종 프로퍼티 수정

기존 Bool 타입으로 뜨겁냐 안뜨겁냐를 가리던 프로퍼티를 온도를 가지는 프로퍼티로 변경하여 내부 메소드에서 특정 기준보다 뜨거운지 안뜨거운지를 리턴하게 함

1Consumption committed:
[기능 수정] Date 비교 메소드 수정

Date 타입을 가공하지 않고 Date 타입끼리 비교하도록 수정

1Consumption commented:
피드백 반영하여 수정 하였습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -16,39 +16,40 @@ class Beverage {
         return balance >= price
     }
     
     func isSafe() -> Bool {
```
한 가지 팁을 주자면 이렇게 바교할 때 기준값도 매개 변수로 넘기도록 만들면 좀 더 유연한 함수가 됩니다

godrm commented on VendingMachineApp/VendingMachineAppTests/VendingMachineTests.swift:
```diff
@@ -35,7 +35,7 @@ class VendingMachineTests: XCTestCase {
     
     func testAddBalance() {
-        XCTAssertEqual(machine.getBalance(), 1000)
+        XCTAssertEqual(machine.balance, 1000)
```
가격을 가져와서 비교하는 방식이 아니라 
뭔가 추가하고 확인할 수 있어야 하는게 아닐까요?
테스트 함수는 모두 독립적으로 개별적으로 실행됩니다. 
테스트는 add를 테스트한다고 써있는데 기능 구현이 없네요

godrm commented:
다 고쳐진게 맞나요. 마지막에 남긴건 다음 단계에서 수정해보세요

1Consumption commented:
마지막 클로저 부분은 VendingMachine을 포함하는 클래스가 생기면 구현하려고 했습니다! 다음 단계에서 고려하겠습니다.

godrm commented:
클로저말고 방금전에 남긴 두 개 얘기였습니다. 개별적으로 대응하지 않으면 리뷰어도 맥락을 놓칠 수 있습니다.
